File: 2024/python/day07.py
import itertools
from functools import lru_cache
from pathlib import Path

from timing_util import time_solution

OPERATIONS = ["+", "*", "||"]


def left_to_right_eval(nums: list[int], ops: tuple[str, ...], target: int) -> int:
    """
    We can evaluate the expression left to right.
    """
    result = nums[0]
    for idx, num in enumerate(nums[1:]):
        if ops[idx] == "+":
            result += num
        elif ops[idx] == "*":
            result *= num
        elif ops[idx] == "||":
            result = int(str(result) + str(num))

        if result > target:
            return result

    return result


def can_find_equalizer(target: int, nums: list[int]) -> bool:
    """Operators are + and *
    We are assuming operator precedence is left to right
    not PEMDAS fine.

    One initial check is if adding all the numbers
    together is equal to the target, then we can return True.
    If it's greater than the target, then we can return False because
    we know that's the minimum we could have.

    We can also check if multiplying all the numbers together is equal
    to the target.
    If it's greater than the target, then we can return False
    because we know that's the minimum we could have.

    ===

    Yeah so this isn't the case because like think about
    8 1 1 1 1 1 1
    the product is actually lower than the sum... so our checks aren't exactly valid.
    """

    # lowest_possible_value = sum(nums)
    # largest_possible_value = math.prod(nums)
    # if lowest_possible_value == target:
    #     return True

    # if lowest_possible_value > target:
    #     return False

    # if largest_possible_value == target:
    #     return True

    # if largest_possible_value < target:
    #     return False

    # We need to try all combinations of + and * between the numbers
    # For n numbers, we need n-1 operators
    # We can use itertools.product to generate all combinations

    # I really should optimize this.... but.... it's kinda the easiest way
    # we could do some better more sophisticated early checking
    # where we could check if the sum of the numbers is less than the target
    # and the product of the numbers is less than the target
    num_operators_needed = len(nums) - 1
    for ops in itertools.product(OPERATIONS, repeat=num_operators_needed):
        if left_to_right_eval(nums, ops, target) == target:
            return True
    return False


@lru_cache(None)
def can_find_equalizer_cache_helper(target: int, nums: list[int]) -> bool:
    num_operators_needed = len(nums) - 1
    for ops in itertools.product(OPERATIONS, repeat=num_operators_needed):
        if left_to_right_eval(nums, ops, target) == target:
            return True
    return False


def can_find_equalizer_cache(target: int, nums: list[int]) -> bool:
    return can_find_equalizer_cache_helper(target, tuple(nums))


def soln(input_file: Path) -> tuple[int, int]:
    """
    We can analyze each line, and basically just see if
    we can hit the targets with the numbers given.

    We're then just trying to sum the valid targets.
    """
    sum_of_valid_targets = 0
    with open(input_file, "r") as f:
        for line in f:
            cleansed_line = line.strip()
            target_str, nums_str = cleansed_line.split(":")
            target = int(target_str)
            nums = list(map(int, nums_str.split()))
            if can_find_equalizer_cache(target, nums):
                sum_of_valid_targets += target

    return sum_of_valid_targets, 0


if __name__ == "__main__":
    curr_dir = Path(__file__).parent
    input_file = curr_dir.parent / "inputs" / "day07.txt"
    # input_file = curr_dir.parent / "inputs" / "day07_small.txt"
    # print(soln(input_file))
    time_solution("day07", soln, input_file)

File: 2024/rust/src/day07.rs
use crate::solution::Solution;
use itertools::Itertools;
use std::collections::HashSet;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

pub struct Day07;

// clone and copy are used to make sure that the enum is copied and not moved
#[derive(Clone, Copy, Debug)]
enum Operation {
    Add,
    Multiply,
    Concatenate,
}

// &[i64] is a slice of i64
// &[Operation] is a slice of Operation
// i64 is the return type
fn left_to_right_eval(nums: &[i64], ops: &[Operation], target: i64) -> i64 {
    let mut result = nums[0];
    for (idx, &num) in nums.iter().enumerate().skip(1) {
        match ops[idx - 1] {
            Operation::Add => result += num,
            Operation::Multiply => result *= num,
            Operation::Concatenate => {
                let concatenated = format!("{}{}", result, num);
                result = concatenated.parse::<i64>().unwrap();
            }
        }
        if result > target {
            return result;
        }
    }
    result
}

fn can_find_equalizer(target: i64, nums: &[i64]) -> bool {
    let num_operators_needed = nums.len() - 1;
    let operations = vec![Operation::Add, Operation::Multiply, Operation::Concatenate];

    // Generate `num_operators_needed` copies of the operations iterator
    let iter = std::iter::repeat(operations).take(num_operators_needed);

    // Use `multi_cartesian_product` to generate combinations
    for ops_combination in iter.multi_cartesian_product() {
        if left_to_right_eval(nums, &ops_combination, target) == target {
            return true;
        }
    }
    false
}

fn can_find_equalizer_cache(
    target: i64,
    nums: &[i64],
    cache: &mut HashSet<(i64, Vec<i64>)>,
) -> bool {
    if cache.contains(&(target, nums.to_vec())) {
        return false;
    }
    let result = can_find_equalizer(target, nums);
    if !result {
        cache.insert((target, nums.to_vec()));
    }
    result
}

fn soln(input_file: &Path) -> (i64, i64) {
    let mut sum_of_valid_targets = 0;
    let file = File::open(input_file).expect("Unable to open file");
    let reader = io::BufReader::new(file);

    let mut cache = HashSet::new();

    for line in reader.lines() {
        let line = line.expect("Unable to read line");
        let cleansed_line = line.trim();
        let parts: Vec<&str> = cleansed_line.split(':').collect();
        let target: i64 = parts[0].parse().unwrap();
        let nums: Vec<i64> = parts[1]
            .split_whitespace()
            .map(|x| x.parse().unwrap())
            .collect();

        if can_find_equalizer_cache(target, &nums, &mut cache) {
            sum_of_valid_targets += target;
        }
    }

    (sum_of_valid_targets, 0)
}

impl Solution for Day07 {
    fn run() {
        println!("Running solution for Day 7!");
        let input_file = Path::new(env!("CARGO_MANIFEST_DIR"))
            .join("..")
            .join("inputs/day07.txt");
        let (part1, part2) = soln(&input_file);
        println!("solution: {}", part1);
    }
}
